Fix drag previews to disappear when cancelling drag in Safari

jesxilin opened the pull request:
#### 3b34f7ad517885f6901fc173cba6445c65ad3f6f
<pre>
Fix drag previews to disappear when cancelling drag in Safari
<a href="https://bugs.webkit.org/show_bug.cgi?id=280854">https://bugs.webkit.org/show_bug.cgi?id=280854</a>
<a href="https://rdar.apple.com/136328215">rdar://136328215</a>

Reviewed by Wenson Hsieh.

Currently, drag previews remain on the screen because we do not remove
all of the previews from the view hierarchy.
This is fixed by having a vector of views to be removed instead of just one view.

* Source/WebKit/UIProcess/ios/DragDropInteractionState.h:
(WebKit::DragDropInteractionState::takePreviewViewsForDragCancel):
(WebKit::DragDropInteractionState::takePreviewViewForDragCancel): Deleted.
* Source/WebKit/UIProcess/ios/DragDropInteractionState.mm:
(WebKit::DragDropInteractionState::previewForCancelling):
(WebKit::DragDropInteractionState::dragAndDropSessionsDidBecomeInactive):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView dragInteraction:item:willAnimateCancelWithAnimator:]):

Canonical link: <a href="https://commits.webkit.org/284657@main">https://commits.webkit.org/284657@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/98c5420090d5841cf59066bfce60bf09ec5b9388

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70094 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49497 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/22857 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/74181 "Hash 98c54200 for PR 34658 does not build (failure)") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/21260 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57299 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/21112 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/74181 "Hash 98c54200 for PR 34658 does not build (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14094 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/73160 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45087 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60452 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/74181 "Hash 98c54200 for PR 34658 does not build (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41747 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/19629 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63671 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18230 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/75898 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/14319 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17470 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/63325 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/14359 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60521 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63260 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15552 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11273 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4887 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45302 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/61 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/46376 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/47650 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/46117 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -10749,16 +10749,20 @@ - (void)dragInteraction:(UIDragInteraction *)interaction item:(UIDragItem *)item
-        });
+        for (auto& previewView : previewViews) {
+            [previewView setAlpha:1];
+            page->dragCancelled();
```
This should happen once, outside of the loop.

whsieh replied:
Nit - from code inspection, it seems safe (but unnecessary) to repeatedly call `dragCancelled()` for each `previewView`.

If possible, we should:

1. Reveal all of the preview views (`-setAlpha:1`)
2. Call `dragCancelled()` once, and then `callAfterNextPresentationUpdate`.
3. In the presentation update callback, remove all of the preview views (`-removeFromSuperview`).

aprotyas commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -10749,16 +10749,20 @@ - (void)dragInteraction:(UIDragInteraction *)interaction item:(UIDragItem *)item
+            page->callAfterNextPresentationUpdate([previewView = WTFMove(previewView), protectedSelf = WTFMove(protectedSelf)] {
+                [previewView removeFromSuperview];
+                protectedSelf->_isAnimatingDragCancel = NO;
+            });
```
You should loop through `previewViews` inside this callback, rather than call this on every iteration.

jesxilin committed:
Fix drag previews to disappear when cancelling drag in Safari
https://bugs.webkit.org/show_bug.cgi?id=280854
rdar://136328215

Reviewed by Wenson Hsieh.

Currently, drag previews remain on the screen because we do not remove
all of the previews from the view hierarchy.
This is fixed by having a vector of views to be removed instead of just one view.

* Source/WebKit/UIProcess/ios/DragDropInteractionState.h:
(WebKit::DragDropInteractionState::takePreviewViewsForDragCancel):
(WebKit::DragDropInteractionState::takePreviewViewForDragCancel): Deleted.
* Source/WebKit/UIProcess/ios/DragDropInteractionState.mm:
(WebKit::DragDropInteractionState::previewForCancelling):
(WebKit::DragDropInteractionState::dragAndDropSessionsDidBecomeInactive):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView dragInteraction:item:willAnimateCancelWithAnimator:]):

Canonical link: https://commits.webkit.org/284657@main
